### PR TITLE
Fix transitive input locking.

### LINKED ISSUE #9143: Flakes don't lock transitive input urls
**Describe the bug**

When trying to override a transitive input's url, the lock file is updated on each evaluation, so that input is never locked.

**Steps To Reproduce**

Uses locked input defined by nixpkgs as expected:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs";
    dream2nix = {
      url = "github:nix-community/dream2nix";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };
}
```

Dream2nix's `nixpkgs` input is updated whenever master changes on github
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs";
    dream2nix = {
      url = "github:nix-community/dream2nix";
      inputs.nixpkgs.url = "github:NixOS/nixpkgs/master";
    };
  };
}
```
The lock file's `original` value for dream2nix/nixpkgs contains the wrong value taken from its lock file, not from the overridden url

**Expected behavior**

**`nix-env --version` output**
nix-env (Nix) 2.18.1

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by vkryachko**
Fixes reproducibility issue described in #9143

# Motivation

Fixes #9143

# Context

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libexpr/flake/flake.cc [hunk 1]
```diff
@@ -425,6 +428,7 @@ LockedFlake lockFlake(
                        ancestors? */
                     auto i = overrides.find(inputPath);
                     bool hasOverride = i != overrides.end();
+                    bool hasCliOverride = explicitCliOverrides.find(inputPath) != explicitCliOverrides.end();
```
Comment (thufschmitt) [inline src/libexpr/flake/flake.cc:431]: (nit)

```suggestion
                    bool hasCliOverride = explicitCliOverrides.contains(inputPath);
```

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Thanks, that looks good.

The whole resolution logic is a bit scary though. Can you add a test for it? Both to make sure that `.follows` now does the right thing and that `--override-input` still works

**Comment by vkryachko** [comment]
@thufschmitt friendly ping.

**Review by thufschmitt — APPROVED** [review]
Sorry for the delay! And thanks for the test. Looks good now

**Pull request merged**